crates/dbsp/src/circuit/metrics.rs

```diff
@@ -57,9 +66,290 @@ pub const COMPACTION_DURATION: &str = "file.compaction_duration";
+    ) {
+        {
+            let c = self.histogram.read().unwrap();
+            if c.contains_key(&metric) {
+                return;
+            }
+        }
+
+        metric.describe(MetricType::Histogram, unit, description);
+
+        {
+            let mut c = self.histogram.write().unwrap();
+            let h = metrics::histogram!(metric.name(), metric.labels.into_labels());
+            c.insert(metric, h);
+        }
+    }
+
+    /// Records this `value` to the [`Histogram`] for the given `metric`.
+    fn record_histogram(&self, metric: DBSPMetric, value: f64) {
+        let h = self.histogram.read().unwrap();
+        if let Some(histogram) = h.get(&metric) {
+            histogram.record(value);
+        }
+    }
+}
+
+enum MetricType {
+    #[allow(unused)]
+    Counter,
+    Gauge,
+    Histogram,
+}
+
+#[derive(Debug, Hash, PartialOrd, PartialEq, Eq, Clone)]
+pub(crate) struct DBSPMetric {
+    /// Name of the metric.
+    key: KeyName,
+
+    /// The global node ID of the operator.
+    global_node_id: GlobalNodeId,
+
+    /// The worker index of the operator.
+    worker_index: usize,
+
+    /// Optional key-value pairs that provide additional metadata about this
+    /// metric.
+    labels: Vec<(MetricString, MetricString)>,
+}
+
+impl DBSPMetric {
+    fn new(
+        key: KeyName,
+        global_node_id: GlobalNodeId,
+        labels: Vec<(MetricString, MetricString)>,
+    ) -> Self {
+        Self {
+            key,
+            global_node_id,
+            worker_index: Runtime::worker_index(),
+            labels,
+        }
+    }
+
+    fn describe(
+        &self,
+        metric_type: MetricType,
+        unit: Option<MetricUnit>,
+        description: Option<String>,
+    ) {
+        let description = description.unwrap_or_default();
+        let key = self.key.clone();
+        match metric_type {
+            MetricType::Counter => {
+                if let Some(unit) = unit {
+                    describe_counter!(key, unit, description);
+                } else {
+                    describe_counter!(key, description);
+                }
+            }
+            MetricType::Gauge => {
+                if let Some(unit) = unit {
+                    describe_gauge!(key, unit, description);
+                } else {
+                    describe_gauge!(key, description);
+                }
+            }
+            MetricType::Histogram => {
+                if let Some(unit) = unit {
+                    describe_histogram!(key, unit, description);
+                } else {
+                    describe_histogram!(key, description);
+                }
+            }
+        }
+    }
+
+    fn name(&self) -> String {
+        format!("dbsp_{}", self.key.as_str(),)
+    }
+}
+
+/// Sets the value for the [`Gauge`] to `value`.
+///
+/// The metric is named in the form: `dbsp_{name}`.
+/// Automatically adds the worker index and global node id as labels.
+pub(crate) fn gauge(
```

Comment:
It looks like we'll need to separate metric creation from metric update. It makes no sense to pass labels each time you update the counter.

Reply:
Removed the labels, do we still need separate methods to declare metrics and update them?

Reply:
I am still in favor of the API we discussed that doesn't involve dynamic allocation or lookups in the critical path. The idea is to create all metrics during initialization and then just use those handles on each metric update.

The last thing we want is our measurement infrastructure interfering with the pipeline. Case in point: we've just discovered that the tracing crate we were using used way too much memory.